File: src/adapters/openai-provider.ts
/**
 * GenericOpenAIProvider — config-driven LLM reviewer.
 * Reads provider config from ~/.anima/providers.d/<name>.yaml (same format ANIMA uses).
 * Resolves ${ENV_VAR} in apiKey automatically.
 * Auto-appends /chat/completions to baseUrl.
 *
 * Usage:
 *   Seal.withLLM(createReviewerFromProvider('minimax'))
 *   Seal.withLLM(createReviewerFromProvider('xiaomi'))
 *   Seal.withLLM(createReviewerFromProvider('gemini'))
 *   Seal.withLLM(new GenericOpenAIProvider({
 *     baseUrl: 'https://api.custom.com/v1',
 *     model: 'my-model',
 *     apiKey: 'sk-...',
 *   }))
 */
import { LLMReviewerAdapter, LLMSignals, SealInput, PartialVerdict } from '../types.ts'
import { readFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'

export interface ProviderConfig {
  baseUrl: string
  model: string
  apiKey: string
  timeoutMs?: number
}

const SYSTEM_PROMPT = `You are Seal, a quality gate in an AI engineering workflow.

Your job is NOT to be polite or creative. Detect semantic issues the deterministic rules could not catch.

Review AI agent output for semantic correctness, subtle logic bugs, and missing requirements.
The deterministic layer already checked: evidence presence, risk keywords, spec headers, test logs.
Your focus: semantic reasoning, prose logic, subtle edge cases, spec compliance.

Return ONLY valid JSON:
{
  "suspected_issues": ["string — exact reasoning error, max 3"],
  "missing_requirements": ["string — spec requirement clearly absent"],
  "possible_edge_cases": ["string — edge cases to consider"],
  "evidence_gaps": ["string — claims needing evidence that slipped through"],
  "risk_guess": "LOW | MEDIUM | HIGH | CRITICAL",
  "confidence": 0.0
}

confidence: 0.0–1.0. Empty arrays if output is correct. No padding.`

function buildMessage(input: SealInput, partial: PartialVerdict): string {
  const parts: string[] = []
  if (input.spec) parts.push(`=== SPEC ===\n${input.spec}`)
  parts.push(`=== OUTPUT (${input.artifact_type}) ===\n${input.output}`)
  const blocking = partial.deterministic_findings.filter(f => f.is_blocking)
  if (blocking.length) {
    parts.push(`=== ALREADY FLAGGED ===\n${blocking.map(f => `- [${f.rule_id ?? f.type}] ${f.evidence}`).join('\n')}`)
  }
  parts.push(`=== PARTIAL VERDICT ===\ntrust_score: ${partial.trust_score}, risk: ${partial.risk_level}`)
  return parts.join('\n\n')
}

function extractJSON(raw: string): string {
  const stripped = raw.replace(/<think>[\s\S]*?<\/think>/g, '').trim()
  const m = stripped.match(/\{[\s\S]*\}/)
  return m ? m[0] : stripped
}

function resolveEnvRef(value: string): string {
  // Resolve ${VAR_NAME} pattern
  return value.replace(/\$\{([^}]+)\}/g, (_, name) => process.env[name] ?? '')
}

/** Load and parse a provider YAML from ~/.anima/providers.d/<name>.yaml */
export function loadProviderConfig(name: string, modelIndex = 0): ProviderConfig {
  const path = join(homedir(), '.anima', 'providers.d', `${name}.yaml`)
  let raw: string
  try {
    raw = readFileSync(path, 'utf-8')
  } catch {
    throw new Error(`Provider config not found: ${path}`)
  }

  // Minimal YAML parser — handles the providers.d format
  const get = (key: string) => {
    const m = raw.match(new RegExp(`^${key}:\\s*(.+)$`, 'm'))
    return m ? m[1].trim().replace(/^["']|["']$/g, '') : ''
  }

  const baseUrl = get('baseUrl')
  const rawKey  = get('apiKey')
  const apiKey  = resolveEnvRef(rawKey)

  // Extract models list (lines like "  - id: MiniMax-M3")
  const models = [...raw.matchAll(/^\s+-\s+id:\s+(.+)$/gm)].map(m => m[1].trim())
  const model = models[modelIndex] ?? models[0] ?? 'default'

  if (!baseUrl) throw new Error(`Provider ${name}: baseUrl missing`)

  return { baseUrl, model, apiKey }
}

export class GenericOpenAIProvider implements LLMReviewerAdapter {
  private config: Required<ProviderConfig>

  constructor(config: ProviderConfig) {
    this.config = { timeoutMs: 30_000, ...config }
  }

  async review(input: SealInput, partial: PartialVerdict): Promise<LLMSignals> {
    const { baseUrl, model, apiKey, timeoutMs } = this.config
    const endpoint = baseUrl.endsWith('/chat/completions')
      ? baseUrl
      : `${baseUrl.replace(/\/$/, '')}/chat/completions`

    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), timeoutMs)

    let response: Response
    try {
      response = await fetch(endpoint, {
        signal: controller.signal,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': apiKey ? `Bearer ${apiKey}` : 'Bearer none',
        },
        body: JSON.stringify({
          model,
          messages: [
            { role: 'system', content: SYSTEM_PROMPT },
            { role: 'user',   content: buildMessage(input, partial) },
          ],
          response_format: { type: 'json_object' },
          temperature: 0.1,
          max_tokens: 1024,
        }),
      })
    } finally {
      clearTimeout(timer)
    }

    if (!response.ok) {
      const body = await response.text().catch(() => '')
      throw new Error(`[${model}] ${response.status}: ${body.slice(0, 200)}`)
    }

    const data = await response.json() as { choices: Array<{ message: { content: string } }> }
    const raw = data.choices?.[0]?.message?.content ?? '{}'

    let signals: Partial<LLMSignals>
    try { signals = JSON.parse(extractJSON(raw)) }
    catch { throw new Error(`[${model}] invalid JSON: ${raw.slice(0, 200)}`) }

    const VALID_RISKS = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']
    return {
      suspected_issues:     Array.isArray(signals.suspected_issues)    ? signals.suspected_issues    : [],
      missing_requirements: Array.isArray(signals.missing_requirements) ? signals.missing_requirements : [],
      possible_edge_cases:  Array.isArray(signals.possible_edge_cases)  ? signals.possible_edge_cases  : [],
      evidence_gaps:        Array.isArray(signals.evidence_gaps)        ? signals.evidence_gaps        : [],
      risk_guess:           VALID_RISKS.includes(signals.risk_guess ?? '') ? signals.risk_guess as LLMSignals['risk_guess'] : 'MEDIUM',
      confidence:           typeof signals.confidence === 'number' ? Math.max(0, Math.min(1, signals.confidence)) : 0.5,
    }
  }
}

/**
 * Create a reviewer from ~/.anima/providers.d/<name>.yaml
 * modelIndex: which model from the models list (default 0 = first/best)
 */
export function createReviewerFromProvider(name: string, modelIndex = 0): GenericOpenAIProvider {
  return new GenericOpenAIProvider(loadProviderConfig(name, modelIndex))
}

File: python/seal_gate/adapters/openai_provider.py
"""
GenericOpenAIProvider — config-driven LLM reviewer.
Reads provider config from ~/.anima/providers.d/<name>.yaml.
Resolves ${ENV_VAR} in apiKey automatically.
Auto-appends /chat/completions to baseUrl.

Usage:
    from seal_gate.adapters.openai_provider import create_reviewer_from_provider, GenericOpenAIProvider
    Seal.with_llm(create_reviewer_from_provider('minimax'))
    Seal.with_llm(create_reviewer_from_provider('xiaomi'))
    Seal.with_llm(GenericOpenAIProvider(base_url='http://localhost:3546/v1', model='local', api_key='none'))
"""
import json
import os
import re
import urllib.request
import urllib.error
from pathlib import Path
from typing import Optional

SYSTEM_PROMPT = """You are Seal, a quality gate in an AI engineering workflow.

Your job is NOT to be polite or creative. Detect semantic issues the deterministic rules could not catch.

Review AI agent output for semantic correctness, subtle logic bugs, and missing requirements.
The deterministic layer already checked: evidence presence, risk keywords, spec headers, test logs.
Your focus: semantic reasoning, prose logic, subtle edge cases, spec compliance.

Return ONLY valid JSON:
{
  "suspected_issues": ["string — exact reasoning error, max 3"],
  "missing_requirements": ["string — spec requirement clearly absent"],
  "possible_edge_cases": ["string — edge cases to consider"],
  "evidence_gaps": ["string — claims needing evidence that slipped through"],
  "risk_guess": "LOW | MEDIUM | HIGH | CRITICAL",
  "confidence": 0.0
}

confidence: 0.0–1.0. Empty arrays if output is correct. No padding."""


def _resolve_env_ref(value: str) -> str:
    return re.sub(r'\$\{([^}]+)\}', lambda m: os.environ.get(m.group(1), ''), value)


def _extract_json(raw: str) -> str:
    stripped = re.sub(r'<think>[\s\S]*?</think>', '', raw).strip()
    m = re.search(r'\{[\s\S]*\}', stripped)
    return m.group(0) if m else stripped


def _build_message(input_data: dict, partial: dict) -> str:
    parts = []
    if input_data.get('spec'):
        parts.append(f"=== SPEC ===\n{input_data['spec']}")
    parts.append(f"=== OUTPUT ({input_data.get('artifact_type', 'unknown')}) ===\n{input_data.get('output', '')}")
    blocking = [f for f in partial.get('deterministic_findings', []) if getattr(f, 'is_blocking', False)]
    if blocking:
        flagged = '\n'.join(f"- [{getattr(f,'rule_id',getattr(f,'type',''))}] {getattr(f,'evidence','')}" for f in blocking)
        parts.append(f"=== ALREADY FLAGGED ===\n{flagged}")
    parts.append(f"=== PARTIAL VERDICT ===\ntrust_score: {partial.get('trust_score',0)}, risk: {partial.get('risk_level','MEDIUM')}")
    return '\n\n'.join(parts)


def load_provider_config(name: str, model_index: int = 0) -> dict:
    """Load ~/.anima/providers.d/<name>.yaml and return {base_url, model, api_key}."""
    path = Path.home() / '.anima' / 'providers.d' / f'{name}.yaml'
    if not path.exists():
        raise FileNotFoundError(f'Provider config not found: {path}')

    raw = path.read_text()

    def get(key: str) -> str:
        m = re.search(rf'^{key}:\s*(.+)$', raw, re.M)
        return m.group(1).strip().strip('"\'') if m else ''

    base_url = get('baseUrl')
    raw_key  = get('apiKey')
    api_key  = _resolve_env_ref(raw_key)

    models = [m.group(1).strip() for m in re.finditer(r'^\s+-\s+id:\s+(.+)$', raw, re.M)]
    model = models[model_index] if models else 'default'

    if not base_url:
        raise ValueError(f'Provider {name}: baseUrl missing')

    return {'base_url': base_url, 'model': model, 'api_key': api_key}


class GenericOpenAIProvider:
    def __init__(self, base_url: str, model: str, api_key: str, timeout: int = 30):
        # Auto-append /chat/completions if not already present
        if base_url.endswith('/chat/completions'):
            self._endpoint = base_url
        else:
            self._endpoint = base_url.rstrip('/') + '/chat/completions'
        self._model   = model
        self._api_key = api_key
        self._timeout = timeout

    def review(self, input_data: dict, partial: dict) -> dict:
        payload = json.dumps({
            'model': self._model,
            'messages': [
                {'role': 'system', 'content': SYSTEM_PROMPT},
                {'role': 'user',   'content': _build_message(input_data, partial)},
            ],
            'response_format': {'type': 'json_object'},
            'temperature': 0.1,
            'max_tokens': 1024,
        }).encode('utf-8')

        req = urllib.request.Request(
            self._endpoint,
            data=payload,
            headers={
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {self._api_key}' if self._api_key else 'Bearer none',
            },
            method='POST',
        )

        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                data = json.loads(resp.read().decode('utf-8'))
        except urllib.error.HTTPError as e:
            raise RuntimeError(f'[{self._model}] {e.code}: {e.read().decode()[:200]}')

        raw = data.get('choices', [{}])[0].get('message', {}).get('content', '{}')

        try:
            signals = json.loads(_extract_json(raw))
        except json.JSONDecodeError:
            raise RuntimeError(f'[{self._model}] invalid JSON: {raw[:200]}')

        valid_risks = {'LOW', 'MEDIUM', 'HIGH', 'CRITICAL'}
        confidence = signals.get('confidence', 0.5)
        if not isinstance(confidence, (int, float)) or isinstance(confidence, bool):
            confidence = 0.5

        return {
            'suspected_issues':     signals.get('suspected_issues', [])    if isinstance(signals.get('suspected_issues'), list)    else [],
            'missing_requirements': signals.get('missing_requirements', []) if isinstance(signals.get('missing_requirements'), list) else [],
            'possible_edge_cases':  signals.get('possible_edge_cases', [])  if isinstance(signals.get('possible_edge_cases'), list)  else [],
            'evidence_gaps':        signals.get('evidence_gaps', [])        if isinstance(signals.get('evidence_gaps'), list)        else [],
            'risk_guess':           signals.get('risk_guess', 'MEDIUM') if signals.get('risk_guess') in valid_risks else 'MEDIUM',
            'confidence':           max(0.0, min(1.0, float(confidence))),
        }


def create_reviewer_from_provider(name: str, model_index: int = 0, timeout: int = 30) -> GenericOpenAIProvider:
    """Load provider from ~/.anima/providers.d/<name>.yaml and return reviewer."""
    cfg = load_provider_config(name, model_index)
    return GenericOpenAIProvider(base_url=cfg['base_url'], model=cfg['model'],
                                 api_key=cfg['api_key'], timeout=timeout)
